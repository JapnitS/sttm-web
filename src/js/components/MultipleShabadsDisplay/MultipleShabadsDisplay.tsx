/* eslint-disable react/prop-types */
import React, { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom';
import { ReactSortable } from 'react-sortablejs'
import { toMultipleShabadsURL } from '@/util/url/to-multiple-shabad-url';
import BarsIcon from '../Icons/Bars';
import { MinusIcon } from '../Icons/CustomIcons';

interface IShabadProps {
  id: number;
  shabadId: number;
  verse: string;
}

interface IMultipleShabadsDisplayProps {
  shabads: IShabadProps[];
  handleClearShabads: () => {};
  removeMultipleShabads: (id: number) => {};
  setMultipleShabads: () => {}
}

const MultipleShabadsDisplay: React.FC<IMultipleShabadsDisplayProps> = ({
  shabads,
  handleClearShabads,
  removeMultipleShabads,
}) => {

  const [sortableState, setSortableState] = useState<IShabadProps[]>(shabads);
  const history = useHistory();

  const onRemove = (e: React.MouseEvent<HTMLElement>) => {
    e.preventDefault();
    const { id } = e.currentTarget.dataset;
    removeMultipleShabads(parseInt(id));
  }

  const handleDisplayShabads = () => {
    const shabadIds = sortableState.map(state => state.shabadId)
    shabadIds.length && history.push(toMultipleShabadsURL({ shabadIds }));
  }

  // Update Local State {sortableState} after shabads get Updated
  useEffect(() => {
    setSortableState(shabads)
  }, [shabads])

  return (
    <div className="multiple-shabads-display">
      <h3>Multiple Shabads Display</h3>

      <ReactSortable tag="ul" list={sortableState} setList={setSortableState}>
        {sortableState?.map(({ verse, id }) => (
          <li key={id}>
            <BarsIcon fill="#fff" />
            <div>{verse}</div>
            <button
              title="Remove"
              data-id={id}
              onClick={onRemove}>-</button>
          </li>
        ))}
      </ReactSortable>

      <div className="multiple-shabads-display--footer">
        <button className="btn btn-secondary" onClick={handleClearShabads}>Clear</button>
        <button className="btn btn-primary" onClick={handleDisplayShabads}>Display</button>
      </div>
    </div>
  )
}

export default MultipleShabadsDisplay;