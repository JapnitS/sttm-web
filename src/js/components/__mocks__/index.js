export * from './shabad';
export * from './multipleShabads';
export * from './store';
