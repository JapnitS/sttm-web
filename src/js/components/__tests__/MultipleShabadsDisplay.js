/* global describe, it, expect, jest */ 
import React from 'react';
import { Provider } from 'react-redux';
import {render, screen} from '@testing-library/react'
import '@testing-library/jest-dom';
import { mockStore } from '../__mocks__/';
//import { clearMultipleShabads, removeMultipleShabads, setMultiViewPanel, setMultipleShabads } from '@/features/actions';

import MultipleShabadsDisplay from "../MultipleShabadsDisplay";

describe('Multiple Shabads Display', () => {
  let store;
  let renderedComponent;
 
  beforeEach(() => {
    store = mockStore({
      multipleShabads: [],
      showMultiViewPanel: true
    });

    store.dispatch = jest.fn();

    renderedComponent = render(
      <Provider store={store}>
        <MultipleShabadsDisplay />
      </Provider>
    );
  });

  it('should render with given state from Redux store', () => {
    const { container } = renderedComponent;
    expect(container).toMatchSnapshot();
  });
  
  it('should render with given state from Redux store', () => {
    expect(screen.getByRole('complementary')).toBeVisible();
  });
})