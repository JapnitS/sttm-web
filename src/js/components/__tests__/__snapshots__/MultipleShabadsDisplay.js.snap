// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiple Shabads Display should render successfully 1`] = `
<div>
  <div
    class="multiple-shabads-display"
    data-testid="multi-view"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="multiple-shabads-display--wrapper enable"
    >
      <button
        class="close toast-notification-close-button"
        title="Close"
      >
        Ã—
      </button>
      <div
        class="multiple-shabads-display--header"
      >
        <h3>
          Multi View
        </h3>
        <p>
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. 
        </p>
      </div>
      <div
        class="notification"
      >
        Please search for shabads and press + button next to the results to add an entry
      </div>
      <ul />
      <div
        class="multiple-shabads-display--footer"
      >
        <button
          class="btn btn-secondary"
        >
          Clear
        </button>
        <button
          class="btn btn-primary"
          disabled=""
        >
          Display
        </button>
      </div>
    </div>
  </div>
</div>
`;
